### DISCUSSION TITLE
How to use with native modules?

### DISCUSSION THREAD

**Discussion by zim32** [preactjs/preact#3803] [Q&A] [answered]
- [x ] Check if updating to the latest Preact version resolves the issue

**Describe the bug**
In docs you say **No build tools route**, but i spend 3 hours trying to make it work with hooks and I give up.

**To Reproduce**
Please provide documentation how to use preact hooks withoud build step. Which files to import and how to use it

**Comment by zim32** [comment]
I tried 

```
import { h, cloneElement } from 'https://unpkg.com/preact@latest?module'
import htm from 'https://unpkg.com/htm?module'
import { useState, useRef, useEffect, useCallback } from 'https://unpkg.com/preact@latest/hooks/dist/hooks.module.js?module'
```

and got 

hooks.module.js?module:1 Uncaught TypeError: Cannot read properties of null (reading '__H')

**Comment by zim32** [comment]
Ok it seems this issue is coming from 2020 year and not properly fixed in two years. Seems like this library is abandoned

**Comment by rschristian** [comment]
We have a pinned issue describing the problem and how one might go about solving it: https://github.com/preactjs/preact/issues/2564#issuecomment-636201423

TLDR? It's a difficult set of issues with Unpkg that aren't really fixable without changing the core of how Unpkg works.

> I tried
>```
>import { h, cloneElement } from 'https://unpkg.com/preact@latest?module'
>import htm from 'https://unpkg.com/htm?module'
>import { useState, useRef, useEffect, useCallback } from 'https://unpkg.com/preact@latest/hooks/dist/hooks.module.js?module'
>```
>and got
>
>hooks.module.js?module:1 Uncaught TypeError: Cannot read properties of null (reading '__H')

Odd, I cannot reproduce that. [Here's my attempt to](https://esm.rschristian.dev/#Ly8gU2ltcGxlIHByZWFjdCBhcHAgYnkgQHJzY2hyaXN0aWFuIGZvciBodHRwczovL2dpdGh1Yi5jb20vcHJlYWN0anMvcHJlYWN0L2Rpc2N1c3Npb25zLzM4MDMKLy8gLS0tLS0tLS0tLS0tLS0tLQoKaW1wb3J0IHsgaCwgcmVuZGVyIH0gZnJvbSAnaHR0cHM6Ly91bnBrZy5jb20vcHJlYWN0QGxhdGVzdD9tb2R1bGUnCmltcG9ydCBodG0gZnJvbSAnaHR0cHM6Ly91bnBrZy5jb20vaHRtP21vZHVsZScKaW1wb3J0IHsgdXNlU3RhdGUsIHVzZVJlZiwgdXNlRWZmZWN0LCB1c2VDYWxsYmFjayB9IGZyb20gJ2h0dHBzOi8vdW5wa2cuY29tL3ByZWFjdEBsYXRlc3QvaG9va3MvZGlzdC9ob29rcy5tb2R1bGUuanM/bW9kdWxlJwppbXBvcnQgeyBjc3MgfSBmcm9tICdodHRwczovL2Nkbi5za3lwYWNrLmRldi9nb29iZXInOwoKY29uc3QgaHRtbCA9IGh0bS5iaW5kKGgpOwoKY29uc3QgcmFuZG9tID0gKCkgPT4gYCR7TWF0aC5yYW5kb20oKSAqIDgwfSVgCmNvbnN0IHN0eWxlID0gY3NzYAogICAgcG9zaXRpb246YWJzb2x1dGU7IGZvbnQtc2l6ZTogMnJlbTsKICAgIGJhY2tncm91bmQ6IHRyYW5zcGFyZW50OyBjb2xvcjogI2Y4ODsKICAgIHdpZHRoOiA1cmVtOyBoZWlnaHQ6IDVyZW07CiAgICBib3JkZXItcmFkaXVzOiA1MCU7CmA7CgpmdW5jdGlvbiBBcHAoKSB7CiAgICBjb25zdCBbc2NvcmUsIHNldFNjb3JlXSA9IHVzZVN0YXRlKDApOwogICAgcmV0dXJuIGh0bWxgCiAgICAgICAgPGJ1dHRvbgogICAgICAgICAgICBjbGFzcz0ke3N0eWxlfQogICAgICAgICAgICBvbkNsaWNrPSR7ZSA9PiBzZXRTY29yZShzY29yZSArIDEpfQogICAgICAgICAgICBzdHlsZT0ke3sgbGVmdDogcmFuZG9tKCksIHRvcDogcmFuZG9tKCkgfX0KICAgICAgICA+CiAgICAgICAgICAgICR7c2NvcmV9CiAgICAgICAgPC9idXR0b24+CiAgICBgOwp9CgpyZW5kZXIoaHRtbGA8JHtBcHB9IC8+YCwgZG9jdW1lbnQuYm9keSk7). I copied what you provided exactly, perhaps you had an issue with a cached version of your code? Not sure.

However, the easiest way to combine `preact`, `preact/hooks`, and `htm` is to use `htm/preact/standalone`. It's a super convenient combination of all 3 in one package! [Here's an example of you could use it](https://esm.rschristian.dev). Hope it helps.

**Comment by zim32** [comment]
My example was simplified just to show imorts. Right now  I can see that hooks are using global shared state, and I was using separate render() calls, my intuition says that if you use multiple render() calls you end up in a mess

**Reply by rschristian** [reply]
> My example was simplified just to show imorts

Sorry to say, but it was simplified to the point of not allowing us to help you then.

> Right now I can see that hooks are using global shared state

No, not quite. Hooks do need to be loaded as a singleton (which is what I'm guessing you're referring to?), but that's not "global shared state" really.

>  I was using separate render() calls, my intuition says that if you use multiple render() calls you end up in a mess

That depends a bit on what you mean by "separate render() calls". [Here's an updated example showing rendering multiple trees](https://esm.rschristian.dev/#Ly8gU2ltcGxlIHByZWFjdCBhcHAgYnkgQHJzY2hyaXN0aWFuCi8vIC0tLS0tLS0tLS0tLS0tLS0KCmltcG9ydCB7IGgsIHJlbmRlciB9IGZyb20gJ2h0dHBzOi8vdW5wa2cuY29tL3ByZWFjdEBsYXRlc3Q/bW9kdWxlJzsKaW1wb3J0IGh0bSBmcm9tICdodHRwczovL3VucGtnLmNvbS9odG0/bW9kdWxlJzsKaW1wb3J0IHsgdXNlU3RhdGUsIHVzZVJlZiwgdXNlRWZmZWN0LCB1c2VDYWxsYmFjayB9IGZyb20gJ2h0dHBzOi8vdW5wa2cuY29tL3ByZWFjdEBsYXRlc3QvaG9va3MvZGlzdC9ob29rcy5tb2R1bGUuanM/bW9kdWxlJzsKaW1wb3J0IHsgY3NzIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvZ29vYmVyJzsKCmNvbnN0IGh0bWwgPSBodG0uYmluZChoKTsKCmNvbnN0IHJhbmRvbSA9ICgpID0+IGAke01hdGgucmFuZG9tKCkgKiA4MH0lYApjb25zdCBzdHlsZSA9IGNzc2AKICAgIHBvc2l0aW9uOmFic29sdXRlOyBmb250LXNpemU6IDJyZW07CiAgICBiYWNrZ3JvdW5kOiB0cmFuc3BhcmVudDsgY29sb3I6ICNmODg7CiAgICB3aWR0aDogNXJlbTsgaGVpZ2h0OiA1cmVtOwogICAgYm9yZGVyLXJhZGl1czogNTAlOwpgOwoKZnVuY3Rpb24gQXBwKCkgewogICAgY29uc3QgW3Njb3JlLCBzZXRTY29yZV0gPSB1c2VTdGF0ZSgwKTsKICAgIHJldHVybiBodG1sYAogICAgICAgIDxidXR0b24KICAgICAgICAgICAgY2xhc3M9JHtzdHlsZX0KICAgICAgICAgICAgb25DbGljaz0ke2UgPT4gc2V0U2NvcmUoc2NvcmUgKyAxKX0KICAgICAgICAgICAgc3R5bGU9JHt7IGxlZnQ6IHJhbmRvbSgpLCB0b3A6IHJhbmRvbSgpIH19CiAgICAgICAgPgogICAgICAgICAgICAke3Njb3JlfQogICAgICAgIDwvYnV0dG9uPgogICAgYDsKfQoKZG9jdW1lbnQuYm9keS5pbnNlcnRBZGphY2VudEhUTUwoJ2JlZm9yZWVuZCcsICc8ZGl2IGlkPSJkaXYxIj48L2Rpdj48ZGl2IGlkPSJkaXYyIj48L2Rpdj4nKTsKcmVuZGVyKGh0bWxgPCR7QXBwfSAvPmAsIGRpdjEpOwpyZW5kZXIoaHRtbGA8JHtBcHB9IC8+YCwgZGl2Mik7).

If you can provide your actual use case I can better help you, else I'm just left guessing.

**Comment by zim32** [comment]
I will try will standalone version

**Comment by zim32** [comment, marked as answer]
Ok. Here is code sandbox example.  Tell me what I am doing wrong?

https://codesandbox.io/s/friendly-fire-ssbdeg?file=/index.html

If you open browser console you will see 

`Cannot read properties of null (reading '__H')`

If you remove useState from TestApp component it will work.

**Reply by JoviDeCroock** [reply]
```js
      render(
        html`<${TestApp} />`,
        document.getElementById("test_components_container")
      );
 ```

> Seems like this library is abandoned

If you want to be rude to people spending their time for free you might want to reevaluate your thinking....

**Reply by zim32** [reply]
> ```js
>       render(
>         html`<${TestApp} />`,
>         document.getElementById("test_components_container")
>       );
> ```
> 
> > Seems like this library is abandoned
> 
> If you want to be rude to people spending their time for free you might want to reevaluate your thinking....

Why this does not work?

```
function TestApp()  {
    const [foo, setFoo] = useState(0);
    return h("p", null, "Hello world");
}

render(TestApp(), document.getElementById("test_components_container"));
```

**Reply by zim32** [reply]
> ```js
>       render(
>         html`<${TestApp} />`,
>         document.getElementById("test_components_container")
>       );
> ```
> 
> > Seems like this library is abandoned
> 
> If you want to be rude to people spending their time for free you might want to reevaluate your thinking....

You will now say that I'm doing somethin wrong, but hey, really, if it works and then you add useState and get some Error which does not explain anything, this is non user friendly at all

**Reply by zim32** [reply]
If useState requires some specific code or context, it will be much much better if it throws some error explaining whats wrong

**Reply by rschristian** [reply]
>Why this does not work?
>```
>function TestApp()  {
>    const [foo, setFoo] = useState(0);
>    return h("p", null, "Hello world");
>}
>
>render(TestApp(), document.getElementById("test_components_container"));
>```

That's covered in our docs, see: https://preactjs.com/guide/v10/api-reference#render

> The first argument must be a valid Virtual DOM Element, which represents either a component or an element. When passing a Component, it's important to let Preact do the instantiation rather than invoking your component directly, which will break in unexpected ways:
>```
>const App = () => <div>foo</div>;
>
>// DON'T: Invoking components directly breaks hooks and update ordering:
>render(App(), rootElement); // ERROR
>render(App, rootElement); // ERROR
>
>// DO: Passing components using h() or JSX allows Preact to render correctly:
>render(h(App), rootElement); // success
>render(<App />, rootElement); // success
>```

**Reply by zim32** [reply]
> > Why this does not work?
> > ```
> > function TestApp()  {
> >    const [foo, setFoo] = useState(0);
> >    return h("p", null, "Hello world");
> > }
> > 
> > render(TestApp(), document.getElementById("test_components_container"));
> > ```
> 
> That's covered in our docs, see: https://preactjs.com/guide/v10/api-reference#render
> 
> > The first argument must be a valid Virtual DOM Element, which represents either a component or an element. When passing a Component, it's important to let Preact do the instantiation rather than invoking your component directly, which will break in unexpected ways:
> > ```
> > const App = () => <div>foo</div>;
> > 
> > // DON'T: Invoking components directly breaks hooks and update ordering:
> > render(App(), rootElement); // ERROR
> > render(App, rootElement); // ERROR
> > 
> > // DO: Passing components using h() or JSX allows Preact to render correctly:
> > render(h(App), rootElement); // success
> > render(<App />, rootElement); // success
> > ```

Thanks. That fixes my issue. I thought I return valid vnode and it should just work. But seems like hooks magic requires some specific context.

**Reply by zim32** [reply]
WHen I see that error, I start googling and found tak about bad CDN, modules, redirects and so on, no one points to problem with render component. So problem was that this error bubbles in several scenarious, which leads to context needed for hooks not properly initialized

**Reply by rschristian** [reply]
>  no one points to problem with render component

I think we could've gotten you an answer much quicker if you just posted the code you were using exactly. That's why it is so important to give good reproductions when opening bugs as we're not seeing what you're seeing. We need that extra context or we're left guessing.
